lib/events/dynamoevents/dynamoevents.go

```diff
@@ -348,57 +351,74 @@ func (l *Log) migrateRFD24WithRetry(ctx context.Context) error {
 // jittered interval until retrying migration again. This allows one server to pull ahead
 // and finish or make significant progress on the migration.
 func (l *Log) migrateRFD24(ctx context.Context) error {
-	hasIndexV1, err := l.indexExists(l.Tablename, indexTimeSearch)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	// Table is already up to date.
-	// We use the existence of the V1 index as a completion flag
-	// for migration. We remove it at the end of the migration which
-	// means it is finished if it doesn't exist.
-	if !hasIndexV1 {
-		return nil
-	}
-
-	hasIndexV2, err := l.indexExists(l.Tablename, indexTimeSearchV2)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	// Table does not have the new index, so we send off an API
-	// request to create it and wait until it's active.
-	if !hasIndexV2 {
-		log.Info("Creating new DynamoDB index...")
-		err = l.createV2GSI(ctx)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-	}
-
 	// Kick off a background task to migrate events, this task is safely
 	// interruptible without breaking data and can be resumed from the
 	// last point so no checkpointing state needs to be managed here.
 	go func() {
 		for {
-			// Migrate events to the new format so that the V2 index can use them.
-			log.Info("Starting event migration to v6.2 format")
-			err := l.migrateDateAttribute(ctx)
-			if err != nil {
-				log.WithError(err).Error("Encountered error migrating events to v6.2 format")
-			} else {
-				// Remove the old index, marking migration as complete
-				log.Info("Removing old DynamoDB index")
-				err = l.removeV1GSI()
+			err := func() error {
+				hasIndexV1, err := l.indexExists(l.Tablename, indexTimeSearch)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				// Table is already up to date.
+				// We use the existence of the V1 index as a completion flag
+				// for migration. We remove it at the end of the migration which
+				// means it is finished if it doesn't exist.
+				if !hasIndexV1 {
+					return nil
+				}
+
+				// Creates the v2 index if it doesn't already exist.
+				// This function takes a lock internally to prevent race conditions.
+				err = l.createV2GSI(ctx)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				// Acquire a lock so that only one auth server attempts to perform the migration at any given time.
+				// If an auth server does in a HA-setup the other auth servers will pick up the migration automatically.
+				err = backend.RunWhileLocked(ctx, l.backend, rfd24MigrationLock, rfd24MigrationLockTTL, func(ctx context.Context) error {
```

Comment:
Nit: what about moving the `createV2GSI` `attribute migration` and `old index removal` operations  in one lock ?

Reply:
`createV2GSI` is in a separate lock for a good reason. If it wasn't, a node could detect it doesn't exist and take the lock to acquire it while another node is doing the same thing. If the longer attribute migration process is under the same lock it won't be able to grab the lock and check until that's done but making it aware that the second index already exists is important since the node will refuse search queries until it knows in the index exists.